examples/hybrid/density_current_2dinvariant_rhoe.jl

```diff
@@ -0,0 +1,348 @@
+        top = Operators.Extrapolate(),
+    )
+
+    dρ .= 0 .* cρ
+
+    cw = If2c.(fw)
+    fuₕ = Ic2f.(cuₕ)
+    cuw = Geometry.Covariant13Vector.(cuₕ) .+ Geometry.Covariant13Vector.(cw)
+
+    ce = @. cρe / cρ
+    cI = @. ce - Φ(z) - (norm(cuw)^2) / 2
+    cT = @. cI / C_v + T_0
+    cp = @. cρ * R_d * cT
+
+    h_tot = @. ce + cp / cρ # Total enthalpy at cell centers
+
+    ### HYPERVISCOSITY
+    # 1) compute hyperviscosity coefficients
+    χe = @. dρe = hwdiv(hgrad(h_tot)) # we store χe in dρe
+    χuₕ = @. duₕ = hwgrad(hdiv(cuₕ))
+
+    Spaces.weighted_dss!(dρe)
+    Spaces.weighted_dss!(duₕ)
+
+    κ₄ = 0.0 # m^4/s
+    @. dρe = -κ₄ * hwdiv(cρ * hgrad(χe))
+    @. duₕ = -κ₄ * (hwgrad(hdiv(χuₕ)))
+
+    # 1) Mass conservation
+    dw .= fw .* 0
+
+    # 1.a) horizontal divergence
+    dρ .-= hdiv.(cρ .* (cuw))
+
+    # 1.b) vertical divergence
+    vdivf2c = Operators.DivergenceF2C(
+        top = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+        bottom = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+    )
+    vdivc2f = Operators.DivergenceC2F(
+        top = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+        bottom = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+    )
+    # we want the total u³ at the boundary to be zero: we can either constrain
+    # both to be zero, or allow one to be non-zero and set the other to be its
+    # negation
+
+    # explicit part
+    dρ .-= vdivf2c.(Ic2f.(cρ .* cuₕ))
+    # implicit part
+    dρ .-= vdivf2c.(Ic2f.(cρ) .* fw)
+
+    # 2) Momentum equation
+
+    # curl term
+    hcurl = Operators.Curl()
+    # effectively a homogeneous Dirichlet condition on u₁ at the boundary
+    vcurlc2f = Operators.CurlC2F(
+        bottom = Operators.SetCurl(Geometry.Contravariant2Vector(0.0)),
+        top = Operators.SetCurl(Geometry.Contravariant2Vector(0.0)),
+    )
+
+    fω¹ = hcurl.(fw)
+    fω¹ .+= vcurlc2f.(cuₕ)
+
+    # cross product
+    # convert to contravariant
+    # these will need to be modified with topography
+    fu¹ =
+        Geometry.Contravariant1Vector.(Geometry.Covariant13Vector.(Ic2f.(cuₕ)),)
+    fu³ = Geometry.Contravariant3Vector.(Geometry.Covariant13Vector.(fw))
+    @. dw -= fω¹ × fu¹ # Covariant3Vector on faces
+    @. duₕ -= If2c(fω¹ × fu³)
+
+
+    @. duₕ -= hgrad(cp) / cρ
+    vgradc2f = Operators.GradientC2F(
+        bottom = Operators.SetGradient(Geometry.Covariant3Vector(0.0)),
+        top = Operators.SetGradient(Geometry.Covariant3Vector(0.0)),
+    )
+    @. dw -= vgradc2f(cp) / Ic2f(cρ)
+
+    cE = @. (norm(cuw)^2) / 2 + Φ(z)
+    @. duₕ -= hgrad(cE)
+    @. dw -= vgradc2f(cE)
+
+    # 3) total energy
+
+    @. dρe -= hdiv(cuw * (cρe + cp))
+    @. dρe -= vdivf2c(fw * Ic2f(cρe + cp))
+    @. dρe -= vdivf2c(Ic2f(cuₕ * (cρe + cp)))
+
+    # Uniform 2nd order diffusion
+    ∂c = Operators.GradientF2C()
+    fρ = @. Ic2f(cρ)
+    κ₂ = 75.0 # m^2/s
+
+    ᶠ∇ᵥuₕ = @. vgradc2f(cuₕ.components.data.:1)
```

Comment:
Does it work if we write it as `ᶠ∇ᵥuₕ = @. vgradc2f(cuₕ)`? If so, this formula generalizes to 3d cases as well.

Reply:
In the Ekman layer column case, (u, v) horizontal velocity are used and computed as https://github.com/CliMA/ClimaCore.jl/blob/main/examples/column/hydrostatic_ekman.jl#L148

Reply:
Since I'm treating these variables as scalars, getting the components seems to be the right way to handle it here ; but we should be able to generalise this to vectors in the future.